File: src/Microsoft.Wpf.Interop.DirectX/D3D11Image.cs
namespace Microsoft.Wpf.Interop.DirectX
{
    using Microsoft.Windows.Media;
    using System;
    using System.Windows;
    using System.Windows.Interop;

    public sealed class D3D11Image : D3DImage, IDisposable
    {
        public static readonly DependencyProperty WindowOwnerHandle =
            DependencyProperty.Register("WindowOwner", typeof(IntPtr), typeof(D3D11Image), new UIPropertyMetadata(IntPtr.Zero, new PropertyChangedCallback(HWNDOwnerChanged)));
        
        public static readonly DependencyProperty OnRenderProperty =
            DependencyProperty.Register("OnRender", typeof(Action<IntPtr>), typeof(D3D11Image), new UIPropertyMetadata(null, new PropertyChangedCallback(RenderChanged)));

        public Action<IntPtr> OnRender
        {
            get { return (Action<IntPtr>)GetValue(OnRenderProperty); }
            set { this.SetValue(OnRenderProperty, value); }
        }

        public IntPtr WindowOwner
        {
            get { return (IntPtr)GetValue(WindowOwnerHandle); }
            set { this.SetValue(WindowOwnerHandle, value); }
        }

        // Interop
        private SurfaceQueueInteropHelper Helper { get; set; }

        public void RequestRender()
        {
            this.EnsureHelper();

            // Don't bother with a call if there's no callback registered.
            if (null != this.OnRender)
            {
                this.Helper.RequestRenderD2D();
            }
        }

        public void SetPixelSize(int pixelWidth, int pixelHeight)
        {
            this.EnsureHelper();
            this.Helper.SetPixelSize((uint)pixelWidth, (uint)pixelHeight);
        }

        #region IDisposable Members
        public void Dispose()
        {
            if (this.Helper != null)
            {
                this.Helper.Dispose();
                this.Helper = null;
            }
        }
        #endregion

        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Reliability", "CA2000:Dispose objects before losing scope", Justification = "Lifetime of the new instance belongs to caller of this method.")]
        protected override System.Windows.Freezable CreateInstanceCore()
        {
            return new D3D11Image();
        }

        #region Callbacks
        private static void HWNDOwnerChanged(DependencyObject sender, DependencyPropertyChangedEventArgs args)
        {
            D3D11Image image = sender as D3D11Image;

            if (image != null)
            {
                if (image.Helper != null)
                {
                    image.Helper.HWND = (IntPtr)args.NewValue;
                }
            }
        }

        private static void RenderChanged(DependencyObject sender, DependencyPropertyChangedEventArgs args)
        {
            D3D11Image image = sender as D3D11Image;

            if (image != null)
            {
                if (image.Helper != null)
                {
                    image.Helper.RenderD2D = (Action<IntPtr>)args.NewValue;
                }
            }
        }
        #endregion Callbacks

        #region Helpers
        private void EnsureHelper()
        {
            if (this.Helper == null)
            {
                this.Helper = new SurfaceQueueInteropHelper();
                this.Helper.HWND = this.WindowOwner;
                this.Helper.D3DImage = this;
                this.Helper.RenderD2D = this.OnRender;
            }
        }
        #endregion Helpers
    }
}

File: src/Microsoft.Wpf.Interop.DirectX/D3D11Image.cpp
#include "D3D11Image.h"

namespace Microsoft {
    namespace Wpf {
        namespace Interop {
            namespace DirectX {

                static D3D11Image::D3D11Image()
                {
                    OnRenderProperty = DependencyProperty::Register("OnRender",
                        Action<IntPtr, bool>::typeid,
                        D3D11Image::typeid,
                        gcnew UIPropertyMetadata(nullptr, gcnew PropertyChangedCallback(&RenderChanged)));

                    WindowOwnerProperty = DependencyProperty::Register("WindowOwner",
                        IntPtr::typeid,
                        D3D11Image::typeid,
                        gcnew UIPropertyMetadata(IntPtr::Zero, gcnew PropertyChangedCallback(&HWNDOwnerChanged)));
                }

                D3D11Image::D3D11Image()
                {
                }

                D3D11Image::~D3D11Image()
                {
                    if (this->Helper != nullptr)
                    {
                        this->Helper->~SurfaceQueueInteropHelper();
                        this->Helper = nullptr;
                    }
                }

                Freezable^ D3D11Image::CreateInstanceCore()
                {
                    return gcnew D3D11Image();
                }

                void D3D11Image::HWNDOwnerChanged(DependencyObject^ sender, DependencyPropertyChangedEventArgs args)
                {
                    D3D11Image^ image = dynamic_cast<D3D11Image^>(sender);

                    if (image != nullptr)
                    {
                        if (image->Helper != nullptr)
                        {
                            image->Helper->HWND = static_cast<IntPtr>(args.NewValue);
                        }
                    }
                }

                void D3D11Image::RenderChanged(DependencyObject^ sender, DependencyPropertyChangedEventArgs args)
                {
                    D3D11Image^ image = dynamic_cast<D3D11Image^>(sender);

                    if (image != nullptr)
                    {
                        if (image->Helper != nullptr)
                        {
                            image->Helper->RenderD2D = static_cast<Action<IntPtr, bool>^>(args.NewValue);
                        }
                    }
                }

                void D3D11Image::EnsureHelper()
                {
                    if (this->Helper == nullptr)
                    {
                        this->Helper = gcnew SurfaceQueueInteropHelper();
                        this->Helper->HWND = this->WindowOwner;
                        this->Helper->D3DImage = this;
                        this->Helper->RenderD2D = this->OnRender;
                    }
                }

                void D3D11Image::RequestRender()
                {
                    this->EnsureHelper();

                    // Don't bother with a call if there's no callback registered.
                    if (nullptr != this->OnRender)
                    {
                        this->Helper->RequestRenderD2D();
                    }
                }

                void D3D11Image::SetPixelSize(int pixelWidth, int pixelHeight)
                {
                    this->EnsureHelper();
                    this->Helper->SetPixelSize(static_cast<UInt32>(pixelWidth), static_cast<UInt32>(pixelHeight));
                }
            }
        }
    }
}